python/pyarrow/dataset/protocol.py

```diff
@@ -0,0 +1,77 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Protocol definitions for pyarrow.dataset
+
+These provide the abstract interface for a dataset. Other libraries may implement
+this interface to expose their data, without having to extend PyArrow's classes.
+
+Applications and libraries that want to consume datasets should accept datasets
+that implement these protocols, rather than requiring the specific
+PyArrow classes.
+"""
+from abc import abstractmethod
+from typing import Iterator, List, Optional, Protocol
+
+from pyarrow.dataset import Expression
+from pyarrow import Table, IntegerArray, RecordBatch, RecordBatchReader, Schema
+
+
+class Scanner(Protocol):
+    @abstractmethod
+    def count_rows(self) -> int:
+        ...
+    
+    @abstractmethod
+    def head(self, num_rows: int) -> Table:
+        ...
+
+    @abstractmethod
+    def take(self, indices: IntegerArray) -> Table:
+        ...
+    
+    @abstractmethod
+    def to_table(self) -> Table:
+        ...
+    
+    @abstractmethod
+    def to_batches(self) -> Iterator[RecordBatch]:
+        ...
+
+    @abstractmethod
+    def to_reader(self) -> RecordBatchReader:
+        ...
+
+
+class Scannable(Protocol):
+    @abstractmethod
+    def scanner(self, columns: Optional[List[str]] = None,
+                filter: Optional[Expression] = None, **kwargs) -> Scanner:
+        ...
+    
+    @abstractmethod
+    def schema(self) -> Schema:
+        ...
+
+
+class Fragment(Scannable):
```

Comment:
I would not make the Fragment scannable, and call the schema [physical_schema](https://github.com/apache/arrow/blob/cd6e2a4d2b9373b942da18b4cc82cb41431764d9/python/pyarrow/parquet/core.py#L2479)

In Iceberg it would be very unlikely that you would apply expressions on a fragment. Except for applying residual filtering, see Google doc, but that should not be exposed in the API.

Reply:
The purpose of making the fragment scannable is allowing them to be used in a distributed scan. So the consumer would create the dataset, get the set of fragments that match the query. Then it would serialize each fragment, distribute them amongst it's worker threads / processes, and then scan each fragment when it can.

This is currently how DataFusion (across threads) and Dask (across processes, I think) scan. See the "Consumers" section of https://docs.google.com/document/d/1r56nt5Un2E7yPrZO9YPknBN4EDtptpx-tqOZReHvq1U/edit#

Do you think that would not be viable for Iceberg?

Reply:
Perhaps "fragment" isn't the right word here.  If this is "something that can be scanned" and it can maybe be scanned in parts then we could go back to some old wording we used to have which is "scan tasks".

At the least, we should probably be clear that a fragment is not necessarily a file.

I do think its useful for parallelization.  However, I agree with @Fokko, the fragments should not be scannable in the same way the dataset is.  It doesn't make sense to provide them with separate filters.  They should essentially have `to_reader` and nothing else.

This does bring up concerns regarding threading and resource contention.  E.g. if datafusion was using pyarrow as a data source and spreading the work across many fragments then it probably wouldn't want those individual fragment scanners using their own threads.

Perhaps we can add `use_threads`?  It's a rather coarse control but may be sufficient.

Reply:
Coming back to this. 

> I would not make the Fragment scannable,

How would you pass a projection then in the distributed case?

I was imagining this sequence:

```python
dataset = ...
projection = ...
filter = ...

fragments = dataset.get_fragments(filter)
for fragment in fragments:
    reader = fragment.scanner(filter, projection).to_reader()
    # use reader...
```

> call the schema [physical_schema]

You're right that fragments currently don't expose the schema. I'll remove that from fragments then. The dataset schema can always be serialized and distributed separately.

> Except for applying residual filtering, see Google doc, but that should not be exposed in the API.

What is "residual filtering"? Why should it not be exposed? My expectation is that if the consumer passed a filter `x < 2` then the scanner should never pass any rows where `x >= 2`. Filters should always be fully applied.

Reply:
Let's consider the following expression: `ts > 2022-05-04T10:00:00`, and assume that the table is partitioned by `days(ts)`:
```
Partitions:
Days: |  2022-05-03  |  2022-05-04  | 2022-05-05  | ...  <-- ts_day
                          |<----- predicate matches -----

      | CANNOT MATCH | MIGHT MATCH  | MUST MATCH  |
                     |<--------------- INCLUSIVE --------
                                    |<---- STRICT -------
```
Then there are three situations:

```
Cannot match: no need to read
Might match: read and check the original predicate (residual)
Must match: read and assume it matches the original predicate
```

This way you can avoid a filter that's not needed because we know of the partition information that is always true.

Reply:
Okay. I think the expectation is that residual filtering needs to be done in this API. The way this is handled in PyArrow's implementation is that each fragment has a "guarantee". So the fragments look like:

```
<fragment 1 guarantee="date(ts) = "2022-05-03">
<fragment 2 guarantee="date(ts) = "2022-05-04">
<fragment 3 guarantee="date(ts) = "2022-05-05">
```

When the consumer calls `fragments = dataset.get_fragments(filter)`, fragment 1 is eliminated, because it cannot match. All the fragments that might match are returned.

The consumer then passes the same filter to the scan of each fragment. When it is passed to fragment 2, the filter is executed because there are some rows it might still filter out. When it is passed to fragment 3, the filter is simplified to `true`, since we know all rows already satisfy it based on the guarantee `date(ts) = "2022-05-05`.

In the end, the rows produced by the dataset scanner must not have any rows that would be removed by the filter.

Does that API make sense now?

Reply:
Yes, it does, thanks for the elaborate answer. Much appreciated.